pkg/expression/builtin_time.go

```diff
@@ -2864,17 +2865,58 @@ func (du *baseDateArithmetical) getIntervalFromString(ctx sessionctx.Context, ar
 	if isNull || err != nil {
 		return "", true, err
 	}
-	// unit "DAY" and "HOUR" has to be specially handled.
-	if toLower := strings.ToLower(unit); toLower == "day" || toLower == "hour" {
-		if strings.ToLower(interval) == "true" {
-			interval = "1"
-		} else if strings.ToLower(interval) == "false" {
+
+	interval, err = du.intervalReformatString(ctx.GetSessionVars().StmtCtx.ErrCtx(), interval, unit)
+	return interval, false, err
+}
+
+func (du *baseDateArithmetical) intervalReformatString(ec errctx.Context, str string, unit string) (interval string, err error) {
+	switch strings.ToUpper(unit) {
+	case "MICROSECOND", "MINUTE", "HOUR", "DAY", "WEEK", "MONTH", "QUARTER", "YEAR":
+		str = strings.TrimSpace(str)
+		// a single unit value has to be specially handled.
```

Comment:
The case of `interval true day` will be convert to `interval 1 day` before this function?

Reply:
The behaviors of handling `interval true day` are the same before/after this PR, the only difference is the behavior to handle the string mode `interval "true"` way.

Before this PR:

```
TiDB root@127.0.0.1:test> select "1900-01-01" + interval "true" day;
+------------------------------------+
| "1900-01-01" + interval "true" day |
+------------------------------------+
| 1900-01-02                         |
+------------------------------------+
```

After this PR:

```
TiDB root@127.0.0.1:test> select "1900-01-01" + interval "true" day;
+------------------------------------+
| "1900-01-01" + interval "true" day |
+------------------------------------+
| 1900-01-01                         |
+------------------------------------+
```

This keeps the same with Mysql8.0